File: nodejs/src/tests/tictactoe.test.ts
const TicTacToe = require("../main/TicTacToe.ts")

playWithPositions = (plays, messages) => {

    let input = jest.fn();
    plays.forEach(position => input.mockImplementationOnce(() => position.line).mockImplementationOnce(() => position.column))

    let output = jest.fn().mockImplementationOnce(message => message)

    new TicTacToe(input, output).start()

    expect(output.mock.calls.length).toBe(messages.length)

    messages.forEach(message => expect(output.mock.calls.some(ticTacToe => ticTacToe.includes(message))).toBe(true))

}

describe("invalid move", () => {

    test("typing letters", () => playWithPositions([{ line: 1, column: "x" },
                                                    { line: 1, column: 1 },
                                                    { line: 3, column: 3 },
                                                    { line: 2, column: 1 },
                                                    { line: 2, column: 2 },
                                                    { line: 3, column: 1 }], ["Invalid position, try again!", "Player X won!!!"]))

    test("typing other characters", () => playWithPositions([{ line: 1, column: "$" },
                                                             { line: 1, column: 1 },
                                                             { line: 3, column: 3 },
                                                             { line: 2, column: 1 },
                                                             { line: 2, column: 2 },
                                                             { line: 3, column: 1 }], ["Invalid position, try again!", "Player X won!!!"]))

})

test("no winner", () => playWithPositions([{ line: 1, column: 1 },
                                           { line: 2, column: 2 },
                                           { line: 2, column: 1 },
                                           { line: 3, column: 1 },
                                           { line: 1, column: 3 },
                                           { line: 1, column: 2 },
                                           { line: 3, column: 2 },
                                           { line: 2, column: 3 },
                                           { line: 3, column: 3 }], ["There was no winner."]))

describe("win column", () => {

    test("Player O won!!!", () => playWithPositions([{ line: 3, column: 3 },
                                                     { line: 1, column: 1 },
                                                     { line: 2, column: 2 },
                                                     { line: 2, column: 1 },
                                                     { line: 1, column: 3 },
                                                     { line: 3, column: 1 }], ["Player O won!!!"]))

    test("Player X won!!!", () => playWithPositions([{ line: 1, column: 1 },
                                                     { line: 3, column: 3 },
                                                     { line: 2, column: 1 },
                                                     { line: 2, column: 2 },
                                                     { line: 3, column: 1 }], ["Player X won!!!"]))

})

describe("win diagonal", () => {

    test("Player O won!!!", () => playWithPositions([{ line: 1, column: 2 },
                                                     { line: 1, column: 1 },
                                                     { line: 1, column: 3 },
                                                     { line: 2, column: 2 },
                                                     { line: 3, column: 2 },
                                                     { line: 3, column: 3 }], ["Player O won!!!"]))

    test("Player X won!!!", () => playWithPositions([{ line: 1, column: 1 },
                                                     { line: 1, column: 2 },
                                                     { line: 2, column: 2 },
                                                     { line: 1, column: 3 },
                                                     { line: 3, column: 3 }], ["Player X won!!!"]))

})

describe("win line", () => {

    test("Player O won!!!", () => playWithPositions([{ line: 2, column: 2 },
                                                     { line: 1, column: 1 },
                                                     { line: 3, column: 3 },
                                                     { line: 1, column: 2 },
                                                     { line: 3, column: 2 },
                                                     { line: 1, column: 3 }], ["Player O won!!!"]))

    test("Player X won!!!", () => playWithPositions([{ line: 1, column: 1 },
                                                     { line: 2, column: 2 },
                                                     { line: 1, column: 2 },
                                                     { line: 3, column: 3 },
                                                     { line: 1, column: 3 }], ["Player X won!!!"]))

})

File: python/src/tests/tictactoe_test.py
import unittest

from io import StringIO
from main.tictactoe import TicTacToe
from unittest.mock import patch

class TicTacToeTest(unittest.TestCase):

    def test_invalid_move(self):
        self.play_with_positions("Player O won!!!", "2","2", "m","1", "1","1", "3","3", "1","2", "3","2", "1","23", "1","3")

    def test_no_winner(self):
        self.play_with_positions("There was no winner.", "1","1", "2","2", "3","1", "2","1", "2","3", "1","3", "3","3", "3","2", "1","2")

    def test_win_column(self):
        self.play_with_positions("Player O won!!!", "1","1", "1","2", "2","1", "2","2", "3","3", "3","2")
        self.play_with_positions("Player X won!!!", "1","1", "1","2", "2","1", "2","2", "3","1")

    def test_win_diagonal(self):
        self.play_with_positions("Player O won!!!", "2","1", "1","1", "3","1", "2","2", "1","2", "3","3")
        self.play_with_positions("Player X won!!!", "1","1", "2","1", "2","2", "3","1", "3","3")

    def test_win_line(self):
        self.play_with_positions("Player O won!!!", "2","2", "1","1", "3","3", "1","2", "3","2", "1","3")
        self.play_with_positions("Player X won!!!", "1","1", "2","2", "1","2", "3","3", "1","3")

    def play_with_positions(self, message_end_game, *positions):
        with patch('builtins.input', side_effect=list(positions)):
            with patch('sys.stdout', new_callable=StringIO) as mock_stdout:
                TicTacToe(lambda message: input(message), lambda message, end=None: print(message, end=end)).start()
                self.assertTrue(message_end_game in mock_stdout.getvalue())


if __name__ == '__main__':
    unittest.main()